proposals/nnnn-actors.md

```diff
@@ -140,6 +140,301 @@ extension BankAccount {
+
+The issue and solutions to it are not new, and have been successfully proven in [Orleans's take on the subject](https://dotnet.github.io/orleans/docs/grains/reentrancy.html). The reason we compare to Orleans here is because it's model is fairly similar to Swift's with regards to modeling actors as reference types that express messages and interactions using async/await.
+
+#### Proposal: Structured concurrency / Task-chain - aware reentrant actors
+
+In addition to the above semantics and fine grained control over reentrancy, we can do better than that, thanks to Swift's built in notion of structured concurrency with tasks and child tasks.
+
+Assuming the non-reentrant actors as just discussed, it is still important to recognize that a frequent and usually quite understandable way of interacting between actors which are simply "conversations" between two or more actors in order fo fulfil some initial request.
+
+Thanks to Swift embracing [Structured Concurrency](https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md) as a core building block of it's concurrency story, we are in good position to do _better_ than just outright banning reentrancy. In Swift, every asynchronous operation is part of a `Task` which encapsulates the general computation taking place, and every asynchronous operation spawned from such task becomes a child Task of the current task. Synchronous calls do not change the current task (in a way, one can think of Tasks as similar to Threads, however they are not directly mapped to one another). Using this core capability in Swift's concurrency model, we are able to make actor calls *task-chain aware* and *allow* such calls to be reentrant.
+
+This resolves both the deadlock we encountered in the `Waiter` and `Kitchen` example in the previous section, and even enables implementing mutually recursive actors. In the following–world's silliest isEven implementation, we can see two actors performing either is even/odd checks, and mutually calling out to each other, because of the structured nature of tasks, and task awareness of the actor runtime, such calls would _not (!)_ deadlock under the task-aware runtime: 
+
+```swift
+// WARNING: Don't actually implement an isOdd/isEven check like this, 
+//          it involves multiple executor hops and is therefore very sub-optimal.
+public actor class OddOddy { 
+  let evan: EvenEvan!
+  
+  func isOdd(n: Int) async -> Bool {
+    if n == 0 { return true }
+    return await evan.isEven(num - 1)
+  }
+}
+
+actor class EvenEvan {
+  let oddy: OddOddy!
+
+  func isEven(n: Int) async -> Bool {
+    if n == 0 { return false }
+    return await oddy.isOdd(num - 1)
```

Comment:
Do I understand correctly that an `async let` would start a new unrelated task that would not be allowed to interleave? If so, that's non-intuitive. I think it should be always correct to refactor an `await` expression into an `async let` that is immediately awaited:

```
    return await oddy.isOdd(num - 1)
```

===>

```
    async let result = oddy.isOdd(num - 1)
    return await result
```

Reply:
an `async let` does not introduce a suspension point, as such it does not offer a chance for interleaving (in any model); the interleaving could only happen at lines of code demarcated by `await` even if we had fully re-entrant actor only.

Reply:
My point is that before the refactoring, we had only one parent task (that was executing both isEven and isOdd functions). After the refactoring, `async let` introduces a task for isOdd. Under the proposed task-aware runtime rules, it seems to me that these two pieces of code would be treated differently.